src/common/http.js

```diff
@@ -28,175 +28,176 @@
+				body.length > options.logBodyLength ? '...' : '';
+		// TODO: make sure below does its job in every API call instance
+		// Don't display password or session id in console
+		logBody = logBody.replace(/password":"[^"]+/, 'password":"********');
+		logBody = logBody.replace(/password=[^&]+/, 'password=********');
+		Zotero.debug(`HTTP ${method} ${url}: ${logBody}`);
 		
 		var xmlhttp = new XMLHttpRequest();
+		xmlhttp.timeout = options.timeout;
+		var promise = this._handleState(xmlhttp);
 		try {
-			xmlhttp.open('GET', url, true);
+			xmlhttp.open(method, url, true);
 
-			for (let header in requestHeaders) {
-				xmlhttp.setRequestHeader(header, requestHeaders[header]);
+			for (let header in options.headers) {
+				xmlhttp.setRequestHeader(header, options.headers[header]);
 			}
-			
-			if(xmlhttp.overrideMimeType && responseCharset) {
-				xmlhttp.overrideMimeType("text/plain; charset=" + responseCharset);
+
+			// Maybe should provide "mimeType" option instead. This is xpcom legacy, where responseCharset
+			// could be controlled manually
+			if (options.responseCharset) {
+				xmlhttp.overrideMimeType("text/plain; charset=" + options.responseCharset);
 			}
 			
-			/** @ignore */
-			xmlhttp.onreadystatechange = function() {
-				_stateChange(xmlhttp, onDone);
-			};
-			xmlhttp.send(null);
+			xmlhttp.send(options.body);
 		} catch(e) {
 			Zotero.logError(e);
-			if(onDone) {
-				window.setTimeout(function() {
-					try {
-						onDone({"status":0, "responseText":""});
-					} catch(e) {
-						Zotero.logError(e);
-						return;
-					}
-				}, 0);
-			}
+			promise.reject(e);
 		}
 		
-		return xmlhttp;
+		return promise.then(function(xmlhttp) {
+			if (options.debug) {
+				Zotero.debug(`HTTP ${xmlhttp.status} response: ${xmlhttp.responseText}`);
+			}	
+			
+			let invalidDefaultStatus = options.successCodes === null &&
+				(xmlhttp.status < 200 || xmlhttp.status >= 300);
+			let invalidStatus = Array.isArray(options.successCodes) && !options.successCodes.includes(xmlhttp.status);
+			if (invalidDefaultStatus || invalidStatus) {
+				throw new Zotero.HTTP.StatusError(xmlhttp);
+			}
+			return xmlhttp;
+		});
 	}
+	/**
+	* Send an HTTP GET request via XMLHTTPRequest
+	*
+	* @deprecated Use {@link Zotero.HTTP.request}
+	* @param {String}			url				URL to request
+	* @param {Function} 		onDone			Callback to be executed upon request completion
+	* @param {String}			responseCharset	
+	* @param {N/A}				cookieSandbox	Not used in Connector
+	* @param {Object}			headers			HTTP headers to include with the request
+	* @return {Boolean} True if the request was sent, or false if the browser is offline
+	*/
+	this.doGet = function(url, onDone, responseCharset, cookieSandbox, headers) {
+		Zotero.debug('Zotero.HTTP.doGet is deprecated. Use Zotero.HTTP.request');
+		this.request('GET', url, {responseCharset, headers})
+		.then(onDone)
+		.catch(function(e) {
+			onDone({status: e.status, responseText: e.responseText});
```

Comment:
I think technically this should be called in the second argument to the previous `then`. Here it would be called on a non-HTTP error if `onDone(xmlhttp)` threw an error in the `then`.